app/workers/chase_references.rb

```diff
@@ -1,75 +1,82 @@
+      type: :referee_reference_request,
+      old_type: :reference_request,
+      include_application_form: true,
+    },
+    {
+      after: 14.days,
+      email: :reference_request_chaser_email,
+      type: :reminder_reference_nudge,
+      include_application_form: true,
+    },
+    {
+      after: 28.days,
+      email: :reference_request_chase_again_email,
+      type: :referee_follow_up_missing_references,
+      old_type: :follow_up_missing_references,
+    },
+  ].freeze
+
+  CANDIDATE_CHASERS = [
+    {
+      after: 9.days,
+      email: :chase_reference,
+      type: :candidate_reference_request,
+      old_type: :reference_request,
+    },
+    {
+      after: 16.days,
+      email: :new_referee_request,
+      type: :reference_replacement,
+      extra_params: { reason: :not_responded },
+    },
+    {
+      after: 30.days,
+      email: :chase_reference_again,
+      type: :candidate_follow_up_missing_references,
+      old_type: :follow_up_missing_references,
+    },
+  ].freeze
+
   def perform
-    send_7_day_chaser!
-    send_14_day_chaser!
-    send_21_day_chaser!
-    send_28_day_chaser!
+    REFEREE_CHASERS.each { |chaser| send_chaser(chaser, mailer: RefereeMailer) }
+    CANDIDATE_CHASERS.each { |chaser| send_chaser(chaser, mailer: CandidateMailer) }
   end
 
 private
 
-  def send_7_day_chaser!
-    references = ApplicationReference
-      .joins(:application_form)
-      .feedback_requested
-      .where(application_forms: { recruitment_cycle_year: ApplicationForm.select('candidate_id').maximum(:recruitment_cycle_year) })
-      .where(['requested_at < ?', TimeLimitConfig.chase_referee_by.days.before(Time.zone.now)])
-      .where.not(id: ChaserSent.reference_request.select(:chased_id))
-
-    references.each do |reference|
-      RefereeMailer.reference_request_chaser_email(reference.application_form, reference).deliver_later
-      CandidateMailer.chase_reference(reference).deliver_later
-      ChaserSent.create!(chased: reference, chaser_type: :reference_request)
-    end
-  end
+  def send_chaser(chaser, mailer:)
+    chase_referee_by = chaser[:after].before(Time.zone.now)
+    chaser_type = chaser[:type].to_s
+    previous_chaser_type = chaser_type.sub(/^candidate_|^referee_/, '')
+    rejected_chased_ids = ChaserSent.where(
+      chaser_type: [chaser_type, previous_chaser_type].uniq,
+    ).select(:chased_id)
 
-  def send_14_day_chaser!
-    references = ApplicationReference
-      .joins(:application_form)
-      .feedback_requested
-      .where(application_forms: { recruitment_cycle_year: ApplicationForm.select('candidate_id').maximum(:recruitment_cycle_year) })
-      .where(['requested_at < ?', TimeLimitConfig.replace_referee_by.days.before(Time.zone.now)])
-      .where.not(id: ChaserSent.reference_replacement.select(:chased_id))
+    references = ApplicationReference.referees_to_chase(chase_referee_by: chase_referee_by, rejected_chased_ids: rejected_chased_ids)
```

Comment:
now we're using ruby 3.1, is there some shorthand you can take advantage of here?

Reply:
Done